### PR TITLE
feat: extensibility for REST APIs

### PR DESCRIPTION
**Pull request by jhockett**
#### Description of changes
WIP

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/override.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// This file is used to override the REST API resource configuration
```
Comment (sachscode) [inline packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/override.ts:1]:
```suggestion
// This file is used to override the REST API resource configuration
import { AmplifyApigwResourceTemplate } from '@aws-amplify/cli-overrides-helper';
```

File: packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "target": "es6",
+    "module": "commonjs",
+    "strict": false,
+    "esModuleInterop": true,
```
[the remark this answers is not in the document]
Comment (sachscode) [reply packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/tsconfig.json:6]: esModuleInterop

**Review by sachscode — COMMENTED** [review]
Mostly nits,
just one issue: 
Api Gateway service name is constructed with a space. This may cause issues in file-name generation and reading with the service-name.

File: packages/amplify-container-hosting/package.json [hunk 3]
```diff
@@ -18,7 +18,7 @@
     "test": "jest --coverage --passWithNoTests"
   },
   "dependencies": {
-    "amplify-category-api": "2.33.2",
+    "@aws-amplify/amplify-category-api": "1.0.0",
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-container-hosting/package.json:21]: We use exact versions for amplify packages - if you have `^` on any internal packages, please fix.

File: packages/amplify-e2e-core/src/categories/api.ts [hunk 4]
```diff
@@ -278,126 +277,134 @@ export function updateAPIWithResolutionStrategyWithModels(cwd: string, settings:
...
-        .wait('Choose a lambda source');
+    if (!isFirstRestApi) {
+      chain.wait('Would you like to add a new path to an existing REST API');
 
-      if (settings.existingLambda) {
+      if (settings.path) {
         chain
-          .send(KEY_DOWN_ARROW)
+          .sendConfirmYes()
+          .wait('Please select the REST API you would want to update')
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-e2e-core/src/categories/api.ts:305]: It does, good callout!

File: packages/amplify-container-hosting/lib/ElasticContainer/index.js [hunk 5]
```diff
@@ -6,7 +6,7 @@ const path = require('path');
 
 const constants = require('../constants');
 
-const { EcsAlbStack, NETWORK_STACK_LOGICAL_ID, DEPLOYMENT_MECHANISM, processDockerConfig } = require('amplify-category-api');
+const { EcsAlbStack, NETWORK_STACK_LOGICAL_ID, DEPLOYMENT_MECHANISM, processDockerConfig } = require('@aws-amplify/amplify-category-api');
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-container-hosting/lib/ElasticContainer/index.js:9]: I agree, but I didn't want to modify the containers package more than necessary for these changes.

File: packages/amplify-cli/package.json [hunk 6]
```diff
@@ -36,7 +36,7 @@
     "@aws-cdk/cloudformation-diff": "~1.124.0",
     "amplify-app": "3.0.16",
     "amplify-category-analytics": "2.21.24",
-    "amplify-category-api": "2.33.2",
+    "@aws-amplify/amplify-category-api": "1.0.0",
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-cli/package.json:39]: Nope, see comment below

File: packages/amplify-cli-core/src/cliConstants.ts [hunk 7]
```diff
@@ -32,11 +32,13 @@ export const AmplifyCategories = {
 };
 
 export const AmplifySupportedService = {
+  APIGW: 'API Gateway',
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-cli-core/src/cliConstants.ts:35]: Discussed offline, this is what the service name is today.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/apigw-input-state.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+    // this.addPolicyResourceNameToPaths(answers.paths);
+    await this.createApigwArtifacts();
+
+    this.context.amplify.updateamplifyMetaAfterResourceUpdate(AmplifyCategories.API, this.resourceName, 'dependsOn', answers.dependsOn);
+    return this.resourceName;
+  };
+
+  public migrateApigwResource = async () => {
+    const deprecatedParametersFileName = 'api-params.json';
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/apigw-input-state.ts:123]: This file is no longer being used. I would prefer to keep this constant localized to the migration function so it's not used elsewhere. What do you think @sachscode?

File: packages/amplify-category-api/src/provider-utils/awscloudformation/index.ts [hunk 9]
```diff
@@ -1,77 +1,103 @@
...
+    case AmplifySupportedService.APIGW:
+      const apigwInputState = ApigwInputState.getInstance(context);
+      return apigwInputState.addApigwResource(serviceWalkthroughPromise, options);
     default:
       return legacyAddResource(serviceWalkthroughPromise, context, category, service, options);
   }
 }
 
-export async function addResource(context, category, service, options) {
+export async function addResource(context: $TSContext, service: string, options) {
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/index.ts:89]: AmplifySupportedService is not a type, it's an object.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-rds-walkthrough.ts [hunk 10]
```diff
@@ -24,9 +25,9 @@ export async function serviceWalkthrough(context: $TSContext, defaultValuesFilen
   let appSyncApi: string;
   const apis = Object.keys(amplifyMeta[category]);
 
-  for (let i = 0; i < apis.length; i += 1) {
-    if (amplifyMeta[category][apis[i]].service === 'AppSync') {
-      appSyncApi = apis[i];
+  for (const api of apis) {
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-rds-walkthrough.ts:28]: It's initialized above, there's no need for this check.

File: packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json [hunk 11]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "overrides-for-root-stack",
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json:2]: What action do you want me to take here?

File: packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/tsconfig.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/tsconfig.json:6]: I copied the file from [here](https://github.com/aws-amplify/amplify-cli/blob/extOverrides2/packages/amplify-category-storage/resources/overrides-resource/DynamoDB/tsconfig.json) for consistency.

It should be fine, it fixes some issues with default imports: https://github.com/aws-amplify/amplify-cli/blob/extOverrides2/packages/amplify-category-storage/resources/overrides-resource/DynamoDB/tsconfig.json

File: packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json:2]: For now change it to `amplify-cli-overrides`

[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json:2]: or you can leave it as we have to consolidate all package.json

File: packages/amplify-category-api/src/commands/api/override.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  } else if (service === AmplifySupportedService.APIGW) {
+    // Migration logic goes in here
+    const apigwInputState = ApigwInputState.getInstance(context, selectedResourceName);
+    if (!apigwInputState.cliInputsFileExists()) {
+      if (await prompter.yesOrNo('File migration required to continue. Do you want to continue?', true)) {
+        await apigwInputState.migrateApigwResource();
+        const stackGenerator = new ApigwStackTransform(context, selectedResourceName);
+        stackGenerator.transform();
+      } else {
+        return;
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/src/commands/api/override.ts:64]: The prompt shows up further down the call stack.

```
➜  oldApis amplify-dev override api
? File migration required to continue. Do you want to continue? (Y/n) › 
```

File: packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-api/resources/awscloudformation/overrides-resource/APIGW/package.json:2]: Leaving for now based on offline discussion. Will be addressed in future PR.

**Review by sachscode — APPROVED** [review]
LGTM

**Review by akshbhu — APPROVED** [review]
LGTM

**Pull request merged**